static/app/views/automations/components/editConnectedMonitors.tsx

```diff
@@ -1,66 +1,236 @@
+  });
   return (
-    <div>
-      {connectedMonitors.length > 0 && (
-        <Fragment>
-          <Heading>{t('Connected Monitors')}</Heading>
-          <ConnectedMonitorsList
-            monitors={connectedMonitors}
-            connectedIds={connectedIds}
-            setConnectedIds={setConnectedIds}
-          />
-        </Fragment>
-      )}
-      <Heading>
-        {connectedMonitors.length > 0 ? t('Other Monitors') : t('All Monitors')}
-      </Heading>
-      <div style={{flexGrow: 1}}>
-        <StyledInputGroup>
-          <InputGroup.LeadingItems disablePointerEvents>
-            <IconSearch color="subText" size="sm" />
-          </InputGroup.LeadingItems>
-          <InputGroup.Input
-            placeholder={t('Search for a monitor or project')}
-            type="text"
-            autoComplete="off"
-          />
-        </StyledInputGroup>
-      </div>
+    <Section title={t('All Monitors')}>
+      <DetectorSearch initialQuery={query} onSearch={setQuery} />
       <ConnectedMonitorsList
-        monitors={unconnectedMonitors}
-        connectedIds={connectedIds}
-        setConnectedIds={setConnectedIds}
+        data-test-id="drawer-all-monitors-list"
+        detectors={monitors}
+        connectedDetectorIds={connectedIds}
+        isLoading={isLoading}
+        isError={isError}
+        toggleConnected={toggleConnected}
+        emptyMessage={t('No monitors found')}
+        numSkeletons={10}
       />
-    </div>
+      <Pagination onCursor={setCursor} pageLinks={getResponseHeader?.('Link')} />
+    </Section>
+  );
+}
+
+function ConnectMonitorsDrawer({
+  initialIds,
+  saveConnectedIds,
+}: {
+  initialIds: Set<string>;
+  saveConnectedIds: (ids: Set<string>) => void;
+}) {
+  const organization = useOrganization();
+  const queryClient = useQueryClient();
+  // Because GlobalDrawer is rendered outside of our form context, we need to duplicate the state here
+  const [connectedIds, setConnectedIds] = useState<Set<string>>(initialIds);
+
+  const toggleConnected = ({detector}: {detector: Detector}) => {
+    const sortedIds = Array.from(connectedIds).sort();
+
+    const oldDetectorsData =
+      getApiQueryData<Detector[]>(
+        queryClient,
+        makeDetectorListQueryKey({
+          orgSlug: organization.slug,
+          ids: sortedIds,
+        })
+      ) ?? [];
+
+    const newDetectors = (
+      oldDetectorsData.some(d => d.id === detector.id)
+        ? oldDetectorsData.filter(d => d.id !== detector.id)
+        : [...oldDetectorsData, detector]
+    )
+      // API will return ID ascending, so this avoids re-ordering
+      .sort((a, b) => a.id.localeCompare(b.id));
+    const newDetectorIds = newDetectors.map(d => d.id);
+
+    // Update the query cache to prevent the list from being fetched anew
+    setApiQueryData<Detector[]>(
+      queryClient,
+      makeDetectorListQueryKey({
+        orgSlug: organization.slug,
+        ids: newDetectorIds,
+      }),
+      newDetectors
+    );
```

Comment:
i think setApiQueryData should give you the old data to mutate

Reply:
This use case is a bit weird since I'm trying to set the query data for a list of IDs that may be completely new (so nothing is in the cache yet). For example if you have 0 connected monitors, then add monitor with ID 1, the connected monitors list will fetch `/detectors?id=1` and I need to set the cache for that before the request is made.